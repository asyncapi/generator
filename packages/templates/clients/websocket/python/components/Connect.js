import { Text } from '@asyncapi/generator-react-sdk';

export function Connect({ title }) {
  return (
    <Text newLines={2} indent={2}>
      {
        `def on_open(self, ws):
    print("Connected to ${title} server")

def on_message(self, ws, message):
    self.handle_message(message)

def on_error(self, ws, error):
    print("WebSocket Error:", error)
    self.handle_error(error)

def on_close(self, ws, close_status_code, close_msg):
    print("Disconnected from ${title}", close_status_code, close_msg)

def connect(self):
    """Establish the connection and start the run_forever loop in a background thread."""
    ssl_opts = {"ca_certs": certifi.where()}
    self.ws_app = websocket.WebSocketApp(
        self.url,
        on_open=self.on_open,
        on_message=self.on_message,
        on_error=self.on_error,
        on_close=self.on_close
    )
    # Run the WebSocketApp's run_forever in a separate thread with multithreading enabled.
    def run():
        while not self._stop_event.is_set():
            try:
                print("Starting WebSocket thread...")
                self.ws_app.run_forever(sslopt=ssl_opts)
            except Exception as e:
                print(f"Exception in WebSocket thread: {e}")  # Print full error details

    thread = threading.Thread(target=run, daemon=True)
    thread.start()`
      }
    </Text>
  );
}
