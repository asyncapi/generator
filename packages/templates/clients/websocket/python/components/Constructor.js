import { Text } from '@asyncapi/generator-react-sdk';

export function Constructor({ serverUrl }) {
  return (
    <Text indent={2} newLines={2}>
      {
        `def __init__(self, url: str = "${serverUrl}"):
    """
    Constructor to initialize the WebSocket client.

    Args:
        url (str, optional): The WebSocket server URL. Use it if the server URL is 
        different from the default one taken from the AsyncAPI document.
    """
    self.url = url
    self.ws_app = None  # Instance of WebSocketApp
    self.message_handlers = []      # Callables for incoming messages
    self.error_handlers = []        # Callables for errors
    self.outgoing_processors = []   # Callables to process outgoing messages
    self._stop_event = threading.Event()`
      }
    </Text>
  );
}
