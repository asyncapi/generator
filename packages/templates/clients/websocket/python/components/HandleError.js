import { Text } from '@asyncapi/generator-react-sdk';

export function HandleError() {
  return (
    <Text newLines={2} indent={2}>
      {
        `def handle_error(self, error):
    """Pass the error to all registered error handlers. Generic log message is printed if no handlers are registered."""
    if len(self.error_handlers) == 0:
      print("\\033[91mError occurred:\\033[0m", error)
    else:
      # Call custom error handlers
      for handler in self.error_handlers:
        handler(error)`
      }
    </Text>
  );
}
