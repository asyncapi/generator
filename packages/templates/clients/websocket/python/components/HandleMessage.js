import { Text } from '@asyncapi/generator-react-sdk';

export function HandleMessage() {
  return (
    <Text newLines={2} indent={2}>
      {
        `def handle_message(self, message):
    """Pass the incoming message to all registered message handlers. """
    if len(self.message_handlers) == 0:
      print("\\033[94mReceived raw message:\\033[0m", message)
    else:
      for handler in self.message_handlers:
        handler(message)`
      }
    </Text>
  );
}
