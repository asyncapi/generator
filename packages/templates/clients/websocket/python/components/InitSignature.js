import { Text } from '@asyncapi/generator-react-sdk';

export function InitSignature({ queryParams, serverUrl }) {
  if (!queryParams) {
    return (
      <Text indent={2} newLines={2}>
        {`def __init__(self, url: str = "${serverUrl}"):`}
      </Text>
    );
  }
  
  const queryParamsArguments = queryParams?.map((param) => {
    const paramName = param[0];
    const paramDefaultValue = param[1];
    const defaultValue = paramDefaultValue ? `"${paramDefaultValue}"` : 'None';
    return `${paramName}: str = ${defaultValue}`;
  }).join(', ');
  
  return (
    <Text indent={2} newLines={2}>
      {`def __init__(self, url: str = "${serverUrl}", ${queryParamsArguments}):`}
    </Text>
  );
}
