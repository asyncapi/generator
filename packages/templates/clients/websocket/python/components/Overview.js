import { Text } from '@asyncapi/generator-react-sdk';

export function OverviewSection({ title, description, version, serverUrl }) {
  return (
    <Text newLines={2}>
      {`# ${title}

## Overview

${description || `A WebSocket client for ${title}.`}
- **Version:** ${version}
- **Server URL:** ${serverUrl}
`}
    </Text>
  );
}

