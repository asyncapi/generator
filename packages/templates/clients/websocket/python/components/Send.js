import { Text } from '@asyncapi/generator-react-sdk';

export function Send({ sendOperations }) {
  if (!sendOperations || sendOperations.length === 0) {
    return null;
  }

  return (
    <Text newLines={2} indent={2}>
      {
        `
@staticmethod
async def _send(message, socket):
    """
    Internal helper to handle the actual sending logic.

    Args:
        message (dict or str): The message to send.
        socket (websockets.WebSocketCommonProtocol): The WebSocket to send through.

    Notes:
        If message is a dictionary, it will be automatically converted to JSON.
    """
    try:
        if isinstance(message, dict):
            message = json.dumps(message)
        await socket.send(message)
    except Exception as e:
        print("Error sending:", e)`
      }
    </Text>
  );
}
