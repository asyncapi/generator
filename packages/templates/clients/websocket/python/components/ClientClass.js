import { Text } from '@asyncapi/generator-react-sdk';
import { Constructor } from './Constructor';
import { Connect } from './Connect';
import { RegisterMessageHandler } from './RegisterMessageHandler';
import { RegisterErrorHandler } from './RegisterErrorHandler';
import { HandleMessage } from './HandleMessage';
import { SendEchoMessage } from './SendEchoMessage';
import { CloseConnection } from './CloseConnection';
import { RegisterOutgoingProcessor } from './RegisterOutgoingProcessor';
import { HandleError } from './HandleError';

export function ClientClass({ clientName, serverUrl, title }) {
  return (
    <Text>
      <Text newLines={2}>
        {`class ${clientName}:`}
      </Text>
      <Constructor serverUrl={serverUrl} />
      <Connect title={title} />
      <RegisterMessageHandler />
      <RegisterErrorHandler />
      <RegisterOutgoingProcessor />
      <HandleMessage />
      <HandleError />
      <SendEchoMessage />
      <CloseConnection />
    </Text>
  );
}
