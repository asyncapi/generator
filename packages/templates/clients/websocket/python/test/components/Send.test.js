import path from 'path';
import { render } from '@asyncapi/generator-react-sdk';
import { Parser, fromFile } from '@asyncapi/parser';
import { Send } from '../../components/Send';

const parser = new Parser();
const asyncapi_websocket_query = path.resolve(__dirname, '../../../../../../helpers/test/__fixtures__/asyncapi-websocket-query.yml');

describe('Testing of Send component', () => {
  let parsedAsyncAPIDocument;

  beforeAll(async () => {
    const parseResult = await fromFile(parser, asyncapi_websocket_query).parse();
    parsedAsyncAPIDocument = parseResult.document;
  });

  test('render Send component with send operations', () => {
    const sendOperations = parsedAsyncAPIDocument.operations().filterBySend();
    const result = render(<Send sendOperations={sendOperations} />);
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });

  test('renders nothing when send operations is empty', () => {
    const result = render(<Send sendOperations={[]} />);
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });
  test('render nothing when send operations is null', () => {
    const result = render(<Send sendOperations={null} />);
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });
  test('renders nothing without send operations', () => {
    const result = render(<Send />);
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });
});