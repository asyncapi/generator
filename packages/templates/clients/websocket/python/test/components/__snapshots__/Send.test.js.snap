// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing of Send component render Send component with send operations 1`] = `
"@staticmethod
  async def _send(message, socket):
      \\"\\"\\"
      Internal helper to handle the actual sending logic.

      Args:
          message (dict or str): The message to send.
          socket (websockets.WebSocketCommonProtocol): The WebSocket to send through.

      Notes:
          If message is a dictionary, it will be automatically converted to JSON.
      \\"\\"\\"
      try:
          if isinstance(message, dict):
              message = json.dumps(message)
          await socket.send(message)
      except Exception as e:
          print(\\"Error sending:\\", e)"
`;

exports[`Testing of Send component render nothing when send operations is null 1`] = `""`;

exports[`Testing of Send component renders nothing when send operations is empty 1`] = `""`;

exports[`Testing of Send component renders nothing without send operations 1`] = `""`;
